---
title: "Single-Cell_clustering_code"
author: "Carolina Segami"
date: "3/03/2023"
output: pdf_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "../Analysis")
```

## Code for clustering single cells of 6 individuals of Dataset 2 of the study.


#Load libraries

```{r libraries}
library(Seurat)
library(SeuratDisk)
#library(SeuratWrappers)
library(ggplot2)
library(cowplot)
library(dplyr)
library(tidyr)
library(ggbeeswarm)
library(data.table)
library(reshape2)
```

## Transforming loom file

We start with a loom file created by velocyto.py from cellranger output plus gtf file. Contains information of unspliced and spliced mRNA. I will write the code in this chunk for one individual/sample but these steps were done for each of the 6 individuals in this Data set.

```{r}
Bird1_dat<- ReadVelocity(file="../Bird1/velocyto/Bird1.loom") #Do this for all individuals, in this case there were 6 individuals

Bird1 <- as.Seurat(x = Bird1_dat)
Bird1[["RNA"]] <- Bird1[["spliced"]]

#Now after all individuals are done, combine in a single seurat object

pure_species<-merge(Bird1, y=c(Bird2,Bird3,Bird4,Bird5,Bird8), 
                    add.cell.ids=c("Bird1","Bird2","Bird3","Bird4",
                                   "Bird5","Bird8"), project="Ficedula")

saveRDS(pure_species, "purespecies.rds")

```

Load seurat objects (already subsetted) or continue

```{r pressure, echo=FALSE}

#Load combined seurat objects

pure<-readRDS("../purespecies.rds")

```

#Pure species
We start with pure species normalization.
```{r normalization}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
pure[["percent.mt"]]<-PercentageFeatureSet(pure, features = c("ENSFALG00000016230","ENSFALG00000016231","ENSFALG00000016232","ENSFALG00000016233",
     "ENSFALG00000016234","ND1","ENSFALG00000016236","ENSFALG00000016237","ENSFALG00000016238",
     "ND2","ENSFALG00000016240","ENSFALG00000016241","ENSFALG00000016242","ENSFALG00000016243",
     "ENSFALG00000016244","COX1","ENSFALG00000016246","ENSFALG00000016247","COX2",
     "ENSFALG00000016249","ATP8","ATP6","COX3","ENSFALG00000016253","ND3","ENSFALG00000016255",
     "ND4L","ND4","ENSFALG00000016258","ENSFALG00000016259","ENSFALG00000016260","ND5","CYTB",
     "ENSFALG00000016263","ENSFALG00000016264","ND6","ENSFALG00000016266")) 

# Visualize QC metrics as a violin plot
VlnPlot(pure, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

plot2 <- FeatureScatter(pure, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 <- FeatureScatter(pure, feature1 = "nCount_RNA", feature2 = "percent.mt")

CombinePlots(plots = list(plot1, plot2))
plot1

#Subset features
pure2 <- subset(pure, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mt < 5) 

###I had to do this because for some reason the info of the individuals was not in orig.ident so had to make a new metadata column called individual to have that info. 
metadata<-pure2@meta.data
metadata<-select(metadata, orig.ident)
metadata<-tibble::rownames_to_column(metadata)
metadataF<-separate(metadata,rowname, c("name","celltag"),sep=":")
pure2<-AddMetaData(pure2,metadataF$name, col.name = "individual")

pure.list <- SplitObject(pure2, split.by = "individual")

Bird1 <- pure.list[["Bird1_Bird1"]]
Bird3 <- pure.list[["Bird3_Bird3"]]
Bird4 <- pure.list[["BBird4_Bird4"]]
Bird5 <- pure.list[["Bird5_Bird5"]]
Bird2 <- pure.list[["Bird2_Bird2"]]
Bird8 <- pure.list[["Bird8_Bird8"]]

Bird1 <- SCTransform(Bird1, vst.flavor = "v2", verbose = FALSE) %>%
    RunPCA(npcs = 30, verbose = FALSE)
Bird2 <- SCTransform(Bird2, vst.flavor = "v2", verbose = FALSE) %>%
    RunPCA(npcs = 30, verbose = FALSE)
Bird3 <- SCTransform(Bird3, vst.flavor = "v2", verbose = FALSE) %>%
    RunPCA(npcs = 30, verbose = FALSE)
Bird4 <- SCTransform(Bird4, vst.flavor = "v2", verbose = FALSE) %>%
    RunPCA(npcs = 30, verbose = FALSE)
Bird5 <- SCTransform(Bird5, vst.flavor = "v2", verbose = FALSE) %>%
    RunPCA(npcs = 30, verbose = FALSE)
Bird8 <- SCTransform(Bird8, vst.flavor = "v2", verbose = FALSE) %>%
    RunPCA(npcs = 30, verbose = FALSE)

#Now integration using pearson residuals

pure.list <- list(Bird1=Bird1,Bird2=Bird2,Bird3=Bird3,Bird4=Bird4,Bird5=Bird5,Bird8=Bird8)

features <- SelectIntegrationFeatures(object.list = pure.list, nfeatures = 4000)
pure.list <- PrepSCTIntegration(object.list = pure.list, anchor.features = features)

pure.anchors <- FindIntegrationAnchors(object.list = pure.list, normalization.method = "SCT",
    anchor.features = features)

pure.combined.sct <- IntegrateData(anchorset = pure.anchors, normalization.method = "SCT")

###Integrated analysis

pure.combined.sct <- RunPCA(pure.combined.sct, verbose = FALSE)
ElbowPlot(pure.combined.sct)
pure.combined.sct <- RunUMAP(pure.combined.sct, reduction = "pca", dims = 1:20, verbose = FALSE, min.dist = 0.7)
pure.combined.sct <- FindNeighbors(pure.combined.sct, reduction = "pca", dims = 1:20)
pure.combined.sct <- FindClusters(pure.combined.sct, resolution = 0.6)

p4 <- DimPlot(pure.combined.sct, reduction = "umap", label = TRUE)
p4 # <4000 features 4000 anchor features min dist 0.7 res 0.6

#save seurat obj
#saveRDS(pure.combined.sct, file = "../DS2020pure.sct.rds")

p6<-DimPlot(pure.combined.sct, reduction = "umap", split.by = "individual")
p6

#count cells

df=data.frame(cluster=Idents(pure.combined.sct), dataset=pure.combined.sct$"individual")
per <- df  %>%  group_by(cluster) %>% 
  dplyr::count(dataset) %>% 
  ungroup()
#### Total cells of cluster 0 = 1577 that are 14.3 of all cells
```

#Continue analysis pure species
Find markers and assign clusters

```{r find}
#pure.combined.sct<-readRDS("../DS2020pure.sct.rds")

pure.markers <- FindAllMarkers(pure.combined.sct, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

pure.top10 <- pure.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
#write.csv(top10,file = "RECLUSTER_minus0/top10fl2minus0reclust_markers.csv")

all.genes <- rownames(pure.combined.sct)
heatmap_pure <- ScaleData(pure.combined.sct, features = all.genes)
Heatmap<-DoHeatmap(heatmap_pure, features = pure.top10$gene)
Heatmap
## Clear that cluster 0 has no good markers

#subset out 0
pure.combined.sct.minus0<-subset(pure.combined.sct, idents = c("0"), invert=T)
Dimminus0<-DimPlot(pure.combined.sct.minus0, reduction = "umap", label = TRUE)
Dimminus0
#Recluster
pure.minus0.rec <- RunPCA(pure.combined.sct.minus0, verbose = FALSE)
ElbowPlot(pure.combined.sct.minus0)
pure.minus0.rec <- RunUMAP(pure.minus0.rec, reduction = "pca", dims = 1:20, 
                           verbose = FALSE,return.model=T,min.dist = 0.7)
pure.minus0.rec <- FindNeighbors(pure.minus0.rec, reduction = "pca", dims = 1:20)
pure.minus0.rec <- FindClusters(pure.minus0.rec, resolution = 0.7)

Dim_0.6<-DimPlot(pure.minus0.rec, reduction = "umap", label = TRUE)
Dim_0.7<-DimPlot(pure.minus0.rec, reduction = "umap", label = TRUE)

Dim_0.6
Dim_0.7


Dimminus0<-DimPlot(pure.minus0.rec, reduction = "umap", label = TRUE)+NoLegend()
Dimminus0
UMAP_pure<-Dimminus0

#saveRDS
saveRDS(pure.minus0.rec,"../DS2020pureminus0.sct.rds")


```


